Populating sub directories correctly in Shared Directory

jatgarg opened the pull request:
## Description

Populating sub directories correctly in Shared Directory.
When we serialize a new directory on attach and record the csn which is -1 for a newly create one, then on load, we load with -1 csn for that. Now there is no create op for this subdir. So when a new subdir is added in that dir and create op comes for that, it does not pass this check below and we don't process that op.
https://github.com/microsoft/FluidFramework/blob/a51cd2afc8fb11e11c083f9f62fa9e69727e6845/packages/dds/map/src/directory.ts#L2150

Commit:
Populating sub directories correctly in Shared Directory

Abe27342 commented:
are you working on a regression test?

jatgarg commented:
> are you working on a regression test?

Yes trying to add a unit test. I think that is what you are asking, right?

Abe27342 commented:
this change seems like it's probably ok (shouldn't make things worse), but based on your description of the repro, it seems like the 'core' problem might be closer to not having consistent handling for sequence numbers at serialization time w.r.t the detached case. It looks like you're doing a similar thing to merge-tree (use 0 for "universal sequence number" -1 for "unacked"). But fixing this at serialization time seems like a better call.

Of course, some documents are already written with -1s we'll probably need your change here as well. But maybe noted for explicit back-compat.

Commit:
add test

jatgarg commented:
> this change seems like it's probably ok (shouldn't make things worse), but based on your description of the repro, it seems like the 'core' problem might be closer to not having consistent handling for sequence numbers at serialization time w.r.t the detached case. It looks like you're doing a similar thing to merge-tree (use 0 for "universal sequence number" and -1 for "unacked"). But fixing this at serialization time seems like a better call.
> 
> Of course, some documents are already written with -1s we'll need your change here as well. But maybe note explicitly that it's for back-compat.

Added test which was failing before. 
I added this check while populating, so that we can maintain back compatibity and we maintain the correct info on serializing representing the actual state.

Commit:
add test

Commit:
add test

Abe27342 commented:
LGTM. I think we should consider some subset of the following items for future improvements (not this PR):

1. Begin serializing 0 rather than -1 for the CSN before attach happens. IMO we should follow a similar pattern to merge-tree, since the treatment of 0/-1 sequence numbers is basically the same there.
2. Refactor usages of 0/-1 sequence numbers into named constants with some doc on their semantics
3. Tweak the fuzzing harness to exercise detached edit flows: AB#4631.

Thanks for the investigation and fix!!
